### PR TITLE
Validation for VPC Name

### PR DESCRIPTION
**Pull request by prateek2408**
Tracked submodules are build/terraform-beta build/terraform-mapper build/terraform build/ansible build/inspec.

This change addresses the need of adding a way to precheck if the given VPC name as per the allowed naming conventions or not. 

part of  https://github.com/hashicorp/terraform-provider-google/issues/9113

If this PR is for Terraform, I acknowledge that I have:

- [X] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [X] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-the-terraform-providers), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [X] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/third_party/terraform/tests) (for handwritten resources or update tests).
- [X] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/main/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [X] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
compute: added validation for name field on `google_compute_network`
```

### CODE REVIEW

**Review by melinath — CHANGES_REQUESTED** [review]
Looks good.

Could you update the release note according to https://github.com/GoogleCloudPlatform/magic-modules/blob/main/.ci/RELEASE_NOTES_GUIDE.md?

**Comment by melinath** [comment]
I think the failures are unrelated, so just the change requests above

File: mmv1/products/compute/terraform.yaml [hunk 1]
```diff
@@ -1512,6 +1512,9 @@ overrides: !ruby/object:Overrides::ResourceOverrides
           If set to `true`, default routes (`0.0.0.0/0`) will be deleted
           immediately after network creation. Defaults to `false`.
     properties:
+      name: !ruby/object:Overrides::Terraform::PropertyOverride
+        validation: !ruby/object:Provider::Terraform::Validation
+          regex: '(?:[a-z](?:[-a-z0-9]{0,61}[a-z0-9])?)'
```
[the remark this answers is not in the document]
Comment (prateek2408) [reply mmv1/products/compute/terraform.yaml:1517]: Sure here is the document which tells about the best practice- 
https://cloud.google.com/architecture/best-practices-vpc-design#naming

However in terms of format I was not able to find a crisp document around that- https://cloud.google.com/vpc/docs/vpc

[the remark this answers is not in the document]
Comment (prateek2408) [reply mmv1/products/compute/terraform.yaml:1517]: Does that solve the purpose ?

[the remark this answers is not in the document]
Comment (melinath) [reply mmv1/products/compute/terraform.yaml:1517]: It looks like this format is documented at https://cloud.google.com/compute/docs/naming-resources#resource-name-format - that would be the link to use in a comment.

**Comment by prateek2408** [comment]
Following the naming convention format as mentioned in - https://cloud.google.com/compute/docs/naming-resources#resource-name-format

File: mmv1/products/compute/terraform.yaml (hunk 1 above)
Comment (rileykarson) [inline mmv1/products/compute/terraform.yaml:1517]: drive-by-review: Should this call the `validateGCPName` function rather than introduce a new, subtly different regex?

Comment (melinath) [reply mmv1/products/compute/terraform.yaml:1517]: I think Prateek copied this based on your comment [here](https://github.com/hashicorp/terraform-provider-google/issues/9113#issuecomment-838839586) - so on the plus side, it's not a new regex :-) but +1 that using [validateGCPName](https://github.com/GoogleCloudPlatform/magic-modules/blob/bcc6331b8c8f79c723c458de2d74e9c733a4a27e/mmv1/third_party/terraform/utils/validation.go#L85) is probably better, assuming that's [compute-specific](https://cloud.google.com/compute/docs/naming-resources#resource-name-format) or there's other documentation showing it applies to this resource?

That would also remove the need to add a comment in the code to link to the documentation for the regex as part of this PR - though it would maybe be a good idea to _add_ a comment to that function as part of this PR.

Comment (rileykarson) [reply mmv1/products/compute/terraform.yaml:1517]: `validateGCPName` is GCE-specific, it got named at a point we pretty much only supported GCE resources. FWIW (and to my point about slightly different regexes) this is a slightly different regex than address- it's missing the ^ and $ to ensure it captures the entire string. I should have pointed to `validateGCPName` in the first place there, my bad on the change in opinion.

> That would also remove the need to add a comment in the code to link to the documentation for the regex as part of this PR - though it would maybe be a good idea to add a comment to that function as part of this PR.

+1

Comment (melinath) [reply mmv1/products/compute/terraform.yaml:1517]: > it's missing the ^ and $ to ensure it captures the entire string

oof, yeah, good catch. 👍

**Review by melinath — CHANGES_REQUESTED** [review]
@prateek2408 re: the above conversation - please make the following changes:

1. Rename `validateGCPName` to `validateGCEName` and add an explanation of the regex that includes a link to the docs for the regex.
2. Add a new `validateGCPName` as a backwards-compatibility shim. It should call `validateGCEName` and it should have a comment explaining why the shim is needed.
3. update this terraform.yaml file to use `validateGCEName` for validation instead of using a regex.

File: mmv1/products/compute/terraform.yaml (hunk 1 above)
Comment (prateek2408) [reply mmv1/products/compute/terraform.yaml]: Made the following changes-

1. Renamed validateGCPName to validateGCEName
2. Inside validateGCEName when error is thrown added link to naming convention https://cloud.google.com/compute/docs/naming-resources#resource-name-format

File: mmv1/third_party/terraform/utils/validation.go [hunk 2]
```diff
@@ -105,7 +105,7 @@ func validateRegexp(re string) schema.SchemaValidateFunc {
 		value := v.(string)
 		if !regexp.MustCompile(re).MatchString(value) {
 			errors = append(errors, fmt.Errorf(
-				"%q (%q) doesn't match regexp %q", k, value, re))
+				"%q (%q) doesn't match regexp %q for more info refer to https://cloud.google.com/compute/docs/naming-resources#resource-name-format", k, value, re))
```
Comment (melinath) [inline mmv1/third_party/terraform/utils/validation.go:108]:
```suggestion
				"%q (%q) doesn't match regexp %q", k, value, re))
```

validateRegexp is used for other regexes as well, so we can't hardcode a specific link like this into the error message. Hypothetically we could allow callers to pass an optional link for more details but that is probably beyond the scope of this ticket.

File: mmv1/third_party/terraform/utils/validation.go [hunk 3]
```diff
@@ -82,7 +82,7 @@ var rfc1918Networks = []string{
 	"192.168.0.0/16",
 }
 
-func validateGCPName(v interface{}, k string) (ws []string, errors []error) {
+func validateGCEName(v interface{}, k string) (ws []string, errors []error) {
```
Comment (melinath) [inline mmv1/third_party/terraform/utils/validation.go:85]: It would be good to have a comment here to document what the source of this regex is. Something like:

```suggestion
// validateGCEName ensures that a field matches the requirements for Compute Engine resource names
// https://cloud.google.com/compute/docs/naming-resources#resource-name-format
func validateGCEName(v interface{}, k string) (ws []string, errors []error) {
```

**Review by melinath — CHANGES_REQUESTED** [review]
It looks like the downstream is failing to compile with the error:

```
google/resource_compute_forwarding_rule.go:190:19: undefined: validateGCPName
```

Looking at the codebase, there are a few references to the old name left in the tpgtools folder. It's a similar situation - yaml override files that reference the name of the function - so it should just be string replacement.

One example: https://github.com/GoogleCloudPlatform/magic-modules/blob/232c61170f8d6760edc4eb4045f81e90832ac5c0/tpgtools/overrides/compute/forwarding_rule.yaml#L110

**Comment by prateek2408** [comment]
@melinath -  Incorporated the suggestions and comments

**Review by melinath — APPROVED** [review]
lgtm

**Comment by melinath** [comment]
fixed release note formatting
